packages/core/src/services/chains/run.ts

```diff
@@ -103,175 +114,63 @@ async function iterate({
   configOverrides?: ConfigOverrides
 }) {
   const prevText = previousResponse?.text
-  const stepValidator = new ChainValidator({ prevText, chain, providersMap })
-  const step = await stepValidator.call().then((r) => r.unwrap())
+  const chainValidator = new ChainValidator({
+    workspace,
+    prevText,
+    chain,
+    providersMap,
+    configOverrides,
+  })
+  const streamConsumer = new ChainStreamConsumer({
+    controller,
+    previousCount,
+    documentLogUuid,
+  })
 
   try {
-    const newMessagesInStep = step.conversation.messages.slice(previousCount)
-    const sentCount = previousCount + newMessagesInStep.length
-    publishStepStartEvent(controller, step, newMessagesInStep, documentLogUuid)
-
+    const step = await chainValidator.call().then((r) => r.unwrap())
     const providerProcessor = new ProviderProcessor({
       source,
       documentLogUuid,
       config: step.config,
       apiProvider: step.provider,
       messages: step.conversation.messages,
+      saveSyncProviderLogs: step.chainCompleted,
     })
+    const { messageCount, stepStartTime } = streamConsumer.start({ step })
     const aiResult = await ai({
-      workspace,
       messages: step.conversation.messages,
       config: step.config,
       provider: step.provider,
-      schema: getSchemaForAI(step, configOverrides),
-      output: getOutputForAI(step, configOverrides),
+      schema: step.schema,
+      output: step.output,
     })
-
-    await streamAIResult(controller, aiResult)
+    const aiStreamConsumedResult = await streamConsumer.consumeAI(aiResult)
     const response = await providerProcessor.call({
       aiResult,
-      saveSyncProviderLogs: step.chainCompleted,
+      streamConsumedResult: aiStreamConsumedResult,
+      startTime: stepStartTime,
     })
 
     if (step.chainCompleted) {
-      await handleCompletedChain(controller, step, response)
-
+      streamConsumer.chainCompleted({ step, response })
       return response
     } else {
-      publishStepCompleteEvent(controller, response)
-
+      streamConsumer.stepCompleted(response)
```

Comment:
stream consumer seems to have maaany responsibilities

Reply:
4 responsibilities. https://github.com/latitude-dev/latitude-llm/pull/345/files#r1789680070

Reply:
You enter that file and everything is stream-related. Way easier to understand and test than going to the `runChain` and having AI + stream + chain validation altogether. I had a hard time understanding all the flow to see where the errors were.

Reply:
Now are 3 responsibilities https://github.com/latitude-dev/latitude-llm/pull/345#discussion_r1789906334